tfx/components/example_gen/utils.py

```diff
@@ -210,83 +214,171 @@ def _glob_to_regex(glob_pattern: Text) -> Text:
-                          split: example_gen_pb2.Input.Split) -> Text:
-  """Retrieves the most recently updated span matching a given split pattern."""
+def _retrieve_latest_span_version(uri: Text,
+                                  split: example_gen_pb2.Input.Split
+                                  ) -> Tuple[Optional[Text], Optional[Text]]:
+  """Retrieves the most recent span and version for a given split pattern.
+
+  If both Span and Version spec occur in the split pattern, searches for and
+  returns both the latest Span and Version. If only Span exists in the split
+  pattern, searches for the latest Span, and Version is returned as None.
+  If Version is present, but not Span, an error is raised. If neither Span
+  nor Version is present, returns both as None.
+
+  Args:
+    uri: The base path from which files will be searched.
+    split: An example_gen_pb2.Input.Split object which contains a split pattern,
+      to be searched on. 
+  Returns:
+    Tuple of two strings, Span (optional) and Version (optional).
+
+  Raises:
+    ValueError: if any of the following occurs:
+      - If either Span or Version spec is occurs in the split pattern
+        more than once.
+      - If Version spec is provided, but Span spec is not present.
+      - If Span or Version found is not an integer.
+      - If a matching cannot be found for split pattern provided.
+  """
+
   split_pattern = os.path.join(uri, split.pattern)
-  if split_pattern.count(SPAN_SPEC) != 1:
-    raise ValueError('Only one {SPAN} is allowed in %s' % split_pattern)
-
-  split_glob_pattern = split_pattern.replace(SPAN_SPEC, '*')
-  logging.info('Glob pattern for split %s: %s', split.name, split_glob_pattern)
-  split_regex_pattern = _glob_to_regex(split_pattern).replace(SPAN_SPEC, '(.*)')
-  logging.info('Regex pattern for split %s: %s', split.name,
-               split_regex_pattern)
-  if re.compile(split_regex_pattern).groups != 1:
-    raise ValueError('Regex should have only one group')
-
-  files = tf.io.gfile.glob(split_glob_pattern)
+
+  split_glob_pattern = split_pattern
+  split_regex_pattern = _glob_to_regex(split_pattern)
+  
   latest_span = None
-  for file_path in files:
-    result = re.search(split_regex_pattern, file_path)
-    if result is None:
-      raise ValueError('Glob pattern does not match regex pattern')
-    try:
-      span = int(result.group(1))
-    except ValueError:
-      raise ValueError('Cannot not find span number from %s based on %s' %
-                       (file_path, split_regex_pattern))
-    if latest_span is None or span >= int(latest_span):
-      # Uses str instead of int because of zero padding digits.
-      latest_span = result.group(1)
-
-  if latest_span is None:
-    raise ValueError('Cannot not find matching for split %s based on %s' %
-                     (split.name, split.pattern))
-  return latest_span
-
-
-def calculate_splits_fingerprint_and_span(
+  latest_version = None
+
+  if SPAN_SPEC in split.pattern:
```

Comment:
consider this, thus the `rest of the code` doesn't have indentation, and you don't need to have the else that too far from if to improve readability , wdyt?
```
if SPAN_SPEC not in split.pattern:
   if VERSION_SPEC in split.pattern
      raise ValueError
   return xxx

rest of the code
```

Reply:
agreed, the elif is very hard to read. fixed